src/memory.rs

```diff
@@ -0,0 +1,48 @@
+const MEMORY_SIZE: usize = 1 << 16;
+
+pub enum MemoryError {
+    OutOfIndex,
+    EmptyOrigin,
+    Overflow,
+}
+
+pub struct Memory {
+    pub cells: [u16; MEMORY_SIZE],
+}
+
+impl Memory {
+    pub fn new() -> Self {
+        Self {
+            cells: [0; MEMORY_SIZE],
+        }
+    }
+
+    pub fn write(&mut self, address: usize, value: u16) -> Result<(), MemoryError> {
+        if let Some(cell) = self.cells.get_mut(address) {
+            *cell = value;
+            Ok(())
+        } else {
+            Err(MemoryError::OutOfIndex)
+        }
+    }
+
+    pub fn read(&self, address: usize) -> Option<u16> {
+        self.cells.get(address).copied()
+    }
+
+    pub fn load_memory(&mut self, data: &[u16]) -> Result<(), MemoryError> {
```

Comment:
I would name this `load` or `load_program` or `load_data`

Reply:
I second this. I think `load_program` is slightly more descriptive. It's pretty subjective though.